open-craft/opencraft
# (BB-1402) Add support for setting up arbitrary subdomain.

When creating an instance there's a few [default subdomains](https://github.com/open-craft/opencraft/blob/master/instance/models/mixins/domain_names.py#L170) that are configured by default. In this context *configured* means the correct [configuration](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/load-balancer-v2/templates/backend.map.ctmpl) [line](https://github.com/open-craft/ansible-playbooks/blob/master/playbooks/roles/load-balancer-v2/templates/haproxy.cfg.ctmpl#L105) in [haproxy](https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#4-use_backend)'s [config file](https://cbonte.github.io/haproxy-dconv/2.0/configuration.html#4.2-http-request%20do-resolve) as well as having [cert-manager](https://github.com/open-craft/cert-manager/blob/master/manage_certs.py#L49) use let's encrypt to create the domain's certificate.
The way this configuration works for default subdomains is that Ocim creates a consul k/v pair with the necessary information (instance and domain list) and there are two different `consul watch` processes running that configure OC's haproxy and run `certbot` when there's a change on consul's side.
This PR leverages that by adding a text area in the instance's admin interface. In this text area a user can add multiple sub-domains (one in each line). Each subdomain is verified to be a subdomain of the main `lms` domain and added to the instance's consul configuration. The rest of the configuration (haproxy and cert-manager) works as expected.

# Testing instructions

1. Checkout this branch and run Ocim and consul locally.
2. Create an instance and query consul. You should see a domain config key like this one:
<img width="1431" alt="Screen Shot 2019-07-05 at 5 57 35 PM" src="https://user-images.githubusercontent.com/4007280/60748177-cdc5c080-9f50-11e9-9f01-a23db4936903.png">
3. Go to Django's admin interface and add multiple sub-domains:
<img width="946" alt="Screen Shot 2019-07-05 at 5 57 03 PM" src="https://user-images.githubusercontent.com/4007280/60748206-141b1f80-9f51-11e9-8a60-2bfc23cd8577.png">
4. Save the updated instance.
5. Check consul for the updated domains list. Should look like this (note the extra sub-domains):
<img width="1436" alt="Screen Shot 2019-07-05 at 6 16 30 PM" src="https://user-images.githubusercontent.com/4007280/60748217-29904980-9f51-11e9-97b5-66aa7a1e0e17.png">
